METRON-939: Upgrade ElasticSearch and Kibana

mmiklavc opened the pull request:
**EDIT** This is ready now (11/30/17)

This PR builds on the great work in https://github.com/apache/metron/pull/619

## Contributor Comments
I wanted to get eyes on this despite some outstanding work left. It was no small undertaking moving 3 major versions, so there's quite a bit to review. In the future I would like to recommend we make feature branches for changes of this magnitude. 

This work upgrades Elasticsearch and Kibana both to 5.6. I'm running a local full dev build now, but assuming no failures this will spin up full dev with everything, including the Kibana dashboard (had to be completely rewritten).

### TODO
- [x] Re-create error dashboard - https://github.com/apache/metron/pull/475
- [x] Fix outstanding test failures
- [x] Another merge with master
- [x] Update ES and Kibana docs
- [x] Document using Curator in place of the pruner tool
- [x] Docs for managing Kibana dashboards (no more pickle file - straight JSON using ES's bulk load API. This was not as easy as you would think it would be.)
- [x] Test with Kerberos


## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?


### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

mraliagha commented:
It might not be completely related to this PR, but since _timestamp is removed in ES 5.x, can we add a specific time of indexing at indexing bolt to capture time of indexing. It is useful for the purpose of benchmarking as well as evaluation of SLA.

cestella commented:
@mraliagha Yeah, absolutely.   This task purposefully leaves out major changes that we could do to improve things (e.g. moving back to `.` separated fields instead of `:`).  The attempt is to touch as few things as possible.  Your ask makes a lot of sense and we should add it as a follow-on task.

ottobackwards commented on dependencies_with_url.csv:
```diff
@@ -310,6 +310,29 @@ org.springframework.security.kerberos:spring-security-kerberos-core:jar:1.0.1.RE
 org.springframework.kafka:spring-kafka:jar:1.1.1.RELEASE:compile,ASLv2,https://github.com/spring-projects/spring-kafka
 ch.hsr:geohash:jar:1.3.0:compile,ASLv2,https://github.com/kungfoo/geohash-java
 org.locationtech.spatial4j:spatial4j:jar:0.6:compile,ASLv2,https://github.com/locationtech/spatial4j
```
can we remove any of the older dependencies?

ottobackwards commented on metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/ELASTICSEARCH/5.6.2/package/scripts/slave.py:
```diff
@@ -48,19 +51,30 @@ def slave():
-    File("{0}/elasticsearch.yml".format(params.conf_dir),
+    elastic_site = params.config['configurations']['elastic-site']
+    path = "{0}/elasticsearch.yml".format(params.conf_dir)
+    Logger.info("Cre")
```
I don't think "Cre" is what you wanted here, typo?

mmiklavc replied:
Hm, adding to my list of cleanup tasks.

ottobackwards commented on metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/KIBANA/5.6.2/package/scripts/kibana_master.py:
```diff
@@ -24,6 +24,7 @@
 
 from ambari_commons.os_check import OSCheck
 from ambari_commons.os_family_impl import OsFamilyFuncImpl, OsFamilyImpl
```
Is the ansible module something we are going to distribute?

mmiklavc replied:
Hm, I'm not sure where that came from. I suspect IntelliJ added that in the middle of my edits.

mmiklavc replied:
Heh, yep. That extra import broke the Kibana install hard. I've made most of your requested changes and will have another commit coming shortly.

ottobackwards commented on metron-platform/elasticsearch-shaded/pom.xml:
```diff
@@ -99,7 +148,15 @@
                               </excludes>
                             </filter>
                           </filters>  
```
Maybe a comment as to why we are doing relocations will help future maintainers

mmiklavc replied:
Yes, absolutely agreed.

ottobackwards commented on metron-platform/metron-data-management/pom.xml:
```diff
@@ -231,11 +231,11 @@
             <artifactId>httpclient</artifactId>
             <version>${httpcore.version}</version>
         </dependency>
```
Why not just remove these?

mmiklavc replied:
Yep, absolutely. More cleanup.

ottobackwards commented on metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/bulk/ElasticsearchImportExport.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.List;
+import java.util.Map;
+import org.apache.metron.common.utils.JSONUtils;
+
```
javadoc

ottobackwards commented on metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java:
```diff
@@ -139,13 +141,17 @@ protected SearchResponse search(SearchRequest searchRequest, QueryBuilder queryB
     searchRequest.getSort().forEach(sortField -> searchSourceBuilder.sort(sortField.getField(), getElasticsearchSortOrder(sortField.getSortOrder())));
     Optional<List<String>> fields = searchRequest.getFields();
     if (fields.isPresent()) {
```
why not just remove?

ottobackwards commented on metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/bulk/ElasticsearchImportExportTest.java:
```diff
@@ -0,0 +1,69 @@
+
+public class ElasticsearchImportExportTest {
+
+
```
Is there a tipping point where this should be a file vs. multiline?

mmiklavc replied:
Yeah, I think you're right. I really like having data inline so that you can see the full context of your tests without having to follow multiple branches to piece together what actually happens. I think it's ok for the system under test, but the tests themselves should be much more intuitive and easy to follow. So I think the right answer here is that I've included a lot of extra noise that is unnecessary for the test. I'm going to prune the "_source" object, which will make this very lean and much easier to read. Really good catch @ottobackwards.

ottobackwards commented:
Are there any required changes to the ansible roles?  I thought we injected the templates from there.  Or has the indexing service injection landed?

mmiklavc commented:
@ottobackwards thanks for the review! The templates themselves changed, but the deployment of those templates is the same. I discovered that with a host change in Kibana, I broke the Ambari Blueprint for running up full dev. I found the issue and am running it up now and should have that committed shortly.

mmiklavc commented:
@ottobackwards Oh, afa the data pruner, I believe we can use Curator now as a built-in mechanism for pruning data. Steps will be added to the documentation.

https://www.elastic.co/guide/en/elasticsearch/client/curator/current/index.html

ottobackwards commented:
Ok, killing my vagrant up then, until your commit

ottobackwards commented:
Can you call the curator stuff out in your checklist above?

mmiklavc commented:
@mraliagha @cestella agreed on timestamp.

ottobackwards commented:
Thanks @mmiklavc, those changes look good.  Running up in full dev

mmiklavc commented:
Ok, looks like 2 more minor issues.
1. params.py missing the kibana_server_host, so it's not making its way to kibana.yml
2. The default index mapping template that ES creates when I index the Kibana configuration is not compatible with Kibana.

I'm done fixing 1 already and have the solution for 2. I'm going to blow away the .kibana index, let Kibana recreate it with the expected template, and then extract that into a template file that's used during the Kibana server install. The only hitch is that this is one more thing that potentially needs to be modified/updated on future upgrades of ES and Kibana.

ottobackwards commented:
I got full_dev failure on connecting to metron web.  Can you post when you think full dev is ready and I'll try again then?

mmiklavc commented:
@ottobackwards You should be good to try it now.

ottobackwards commented:
Ran up in full dev, everything running normally, kibana dashboard has data.  Is the color scheme supposed to be so 'miami vice'?

mmiklavc commented:
Whew, that's great news!

Hahahahaha, that's the default colors from the Kibana widgets... I usually refrain from bringing it up, but I'm partially colorblind (even though I used to do a fair amount of Photoshop graphic design work on the side, but it's a bit more effort for me). Anyhow, Kibana had no easy way to pick color names or supply RGB values, so I punted on it for now because I figured that's an easy change. If you would be willing to take a screenshot or the xy coordinate of colors in the color picker that you'd prefer, I'd be more than happy to change it.

ottobackwards commented:
The only issue with it, is if we did that in the old version, then this would be a regression.  Other than that, I'm just waiting to review the docs ;)

ottobackwards commented:
6.0 was released, should we consider going to that while we are at it?

mmiklavc commented:
I was looking at that today as well, and I think that should be a follow-on
considering how large the change is moving from 2.x to 5.x. The changes
I've implemented (and Casey and Nick) for 5.x should make moving to 6.x
much easier once we're ready for it, but we should probably stabilize
master on this before another upgrade. Also, we have a number of fixes to
meta alerts that I need to merge with this branch, which will also take
some time. There are also some enhancements to the dashboard that we're
tossing around as well.

On Nov 14, 2017 6:46 PM, "ottobackwards" <notifications@github.com> wrote:

> 6.0 was released, should we consider going to that while we are at it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/metron/pull/840#issuecomment-344459396>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAoMCxHosc6Ndi3z6K68IWCZWRA71tuuks5s2kJZgaJpZM4QZFWa>
> .
>

mmiklavc commented:
Modified the Kibana chart colors per @ottobackwards request. I don't know if Kibana ever offered fine-grained control over the bar chart colors, but they do not offer it now. This is as close as I was able to get.

![image](https://user-images.githubusercontent.com/658443/33148607-b5d7236c-cf89-11e7-8536-7aebd70b9a4b.png)

mraliagha commented:
Is this the best time to ask for changing field name convention to avoid dot or colon? We are externally using Hive external tables on HDFS data, due to Hive limitations we need to change the Metron field convention. I heard there is a long term plan in future to use ORC files instead of JSON and maybe Hive table can be supported directly. If this is right maybe this is the best time we can move towards changing field seperators accordingly.

mraliagha commented:
One of the issues that we had with the previous mpack was the lack of config segregation for Elasticsearch Master Nodes and Data Nodes. Hence, we have ended up hardcoding specific configuration outside Ambari. With ES 5 it will be even more complex due to the existence of Ingestion Nodes and ML Nodes. Therefore, it would be really nice if we can have multiple tabs in Ambari ES service. For example, one tab for generic configurations and another tab per each node type to segregate specific configurations from generic ones.

mmiklavc commented:
@mraliagha I do think we should consider revisiting the field name conventions, but I'd push for that as a follow-on task. As discussed in other points on this thread, e.g. going straight to ES 6.x, this PR is already massive in scope, and adding further non-critical improvements will increase the surface area of those changes. 

Now that we have the meta alerts fixes in master, I strongly encourage the community to stabilize around as minimal an ES upgrade as possible with incremental improvements and feature enhancements once we've been able to convince ourselves through the unit and integration test suites and functional testing that we're in a healthy and stable place with minimal risk of adding regressions.

mmiklavc commented:
For reference, here is a list of some of the follow-on work we should consider:

- Improvements to Kibana dashboard
- Add new timestamp field to parsers and index templates to take place of _timestamp
- Fix Log4j logging problem - classpath issues
- Upgrade to ES 6.x
- Migrate to new ES REST client
- Improved config management for ES 5.x+ in Ambari
- New field name conventions?

justinleet commented on metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/files/snort_index.template:
```diff
@@ -102,13 +94,25 @@
+            "match_mapping_type": "*"
+          }
         }
       },
```
This brace is extraneous, I'm guessing a merge broke it. Drop it, but keep the comma and we should be good.  I'd just get the formatting lined back up while you're in there.

mraliagha commented:
Yes, I agree. It completely makes sense to minimize the scope and work on stabilizing this version at this moment.

mmiklavc commented:
Ok, the Metron error dashboard is in now. I'll add some additional testing instructions tomorrow, but this PR should be ready for some more vigorous testing. The most easily accessible e2e place to start is with the Kibana dashboards. Checking Kibana minimally proves that data is flowing through the system with the new versions of Elasticsearch and Kibana. The default main dashboard should come up immediately with Bro and Snort data only. 

I recommend first stopping some services:

```
service monit stop
storm kill profiler
```

Yaf is not enabled by default, but you can start the Yaf topology and turn on all sensor stubs by doing the following:

```
# start yaf parser topology
export METRON_HOST=node1
export ZOOKEEPER=${METRON_HOST}:2181
export BROKERLIST=${METRON_HOST}:6667
export METRON_VERSION=0.4.2
export METRON_HOME=/usr/metron/${METRON_VERSION}
$METRON_HOME/bin/start_parser_topology.sh -k $BROKERLIST -z $ZOOKEEPER -s yaf
# start the yaf sensor stub
service sensor-stubs start yaf
```

Now publish some intentionally bad data to  each of the topics so we can check the error dashboard.

```
# publish error data on a cycle
cycleval=1; 
while true; 
do 
    if [ $cycleval -ne 0 ]; 
    then 
        datestamp=$(date "+%s");
        cycleval=0;
    else
        cycleval=1;
    fi;
    echo "bro-garbage-" $datestamp | /usr/hdp/current/kafka-broker/bin/kafka-console-producer.sh --broker-list $BROKERLIST --topic bro;
    echo "snort-garbage-" $datestamp | /usr/hdp/current/kafka-broker/bin/kafka-console-producer.sh --broker-list $BROKERLIST --topic snort;
    echo "yaf-garbage-" $datestamp | /usr/hdp/current/kafka-broker/bin/kafka-console-producer.sh --broker-list $BROKERLIST --topic yaf;
    sleep 2;
done;
```

mmiklavc commented:
For reference, I've added some PDF snapshots of the dashboards to the Jira

- https://issues.apache.org/jira/secure/attachment/12899952/Metron-Dashboard%20-%20Kibana.pdf
- https://issues.apache.org/jira/secure/attachment/12899951/Metron-Error-Dashboard%20-%20Kibana.pdf

mmiklavc commented:
Ok, here is the remaining bit of the test plan for verifying ES. Thanks @cestella  for pulling together some of the public test scripts around this.

# Test Script

Testing Kibana dashboards - see [here](https://github.com/apache/metron/pull/840#issuecomment-348085037)

Testing Instructions beyond the normal smoke test (i.e. letting data
flow through to the indices and checking them).

# Preliminaries

Set an environment variable to indicate `METRON_HOME`:
* `export METRON_HOME=/usr/metron/0.4.2` 

# Deploy the dummy parser
* Edit `$METRON_HOME/config/zookeeper/parsers/dummy.json`:
```
{
  "parserClassName":"org.apache.metron.parsers.json.JSONMapParser",
  "sensorTopic":"dummy"
}
```
* Create the dummy kafka topic:
  `/usr/hdp/current/kafka-broker/bin/kafka-topics.sh --zookeeper node1:2181 --create --topic dummy --partitions 1 --replication-factor 1`
* Persist config changes: `$METRON_HOME/bin/zk_load_configs.sh -m PUSH -i $METRON_HOME/config/zookeeper -z node1:2181`
* Start via `$METRON_HOME/bin/start_parser_topology.sh -k node1:6667 -z node1:2181 -s dummy`

# Send dummy data through
* Edit `~/msg.json` with the following content:
```
{ "guid" : "guid0", "sensor.type" : "dummy", "timestamp" : 100 }
```
* Send `msg.json` through to kafka via `cat ~/msg.json | /usr/hdp/current/kafka-broker/bin/kafka-console-producer.sh --broker-list node1:6667 --topic dummy`
* Validate data has been written to the index:
```
curl -XPOST 'http://localhost:9200/dummy*/_search?pretty' 
```

## Test Case: Update via patch
* Patch the message in ES and create a new field 'project' by executing
  the following:
```
curl -u user:password -X PATCH --header 'Content-Type: application/json' --header 'Accept: */*' -d '{
  "guid" : "guid0",
    "sensorType" : "dummy",
    "patch" : [
    {
      "op": "add"
        , "path": "/project"
        , "value": "metron"
    }
  ]
}' 'http://node1:8082/api/v1/update/patch'
```
* Validate that the message has a field 'project':
```
curl -XPOST 'http://localhost:9200/dummy*/_search?pretty' -d '
{
  "_source" : [ "project" ]
}
'
```

## Test Case: Update via replace 
* Replace the message in ES and create a couple of modifications:
  * new field `new_field` == "brand new"
  * modified `timestamp` == 7
Execute the following:
```
curl -u user:password -X POST --header 'Content-Type: application/json' --header 'Accept: */*' -d '{
     "guid" : "guid0",
     "sensorType" : "dummy",
     "replacement" : {
       "source:type": "dummy",
       "guid" : "guid0",
       "new_field" : "brand new",
       "timestamp" : 7
      }
   }' 'http://node1:8082/api/v1/update/replace'
```
* Validate that the message has a field 'new_field':
```
curl -XPOST 'http://localhost:9200/dummy*/_search?pretty' -d '
{
  "_source" : [ "new_field", "timestamp" ]
}
'
```
## Meta Alerts Test

### Set Up Base Data
We're going to set up a bit of base data.
Retrieve the current list of indices so we know where to put our data
```
curl 'node1:9200/_cat/indices?v'
health status index                     pri rep docs.count docs.deleted store.size pri.store.size
green  open   snort_index_2017.09.06.14   1   0        130            0    180.9kb        180.9kb
green  open   bro_index_2017.09.06.14     1   0        160            0    564.3kb        564.3kb
green  open   .kibana                     1   0         52            0     71.2kb         71.2kb
green  open   metaalert_index             1   0          6            0     62.3kb         62.3kb
```
In this case, we care about `snort_index_2017.09.06.14` and `metaalert_index`.  To make our lives easier, we'll add a couple of stripped down messages to the our snort index (Make sure to sub in the correct index name):
```
curl -XPUT 'node1:9200/snort_index_2017.09.06.14/snort_doc/snort_test_1?pretty' -H 'Content-Type: application/json' -d'
{
  "msg": "snort test alert",
  "ip_dst_port": "8080",
  "ethsrc": "0A:00:27:00:00:00",
  "protocol": "TCP",
  "source:type": "snort",
  "ip_dst_addr": "192.168.66.121",
  "ip_src_addr": "192.168.66.1",
  "threat:triage:rules:0:score": 10,
  "timestamp": 1504708744000,
  "threat:triage:rules:0:reason": null,
  "threat:triage:score": 10,
  "is_alert": "true",
  "ip_src_port": "50187",
  "guid": "snort_test_1"
}
'
```
```
curl -XPUT 'node1:9200/snort_index_2017.09.06.14/snort_doc/snort_test_2?pretty' -H 'Content-Type: application/json' -d'
{
  "msg": "snort test alert 2",
  "ip_dst_port": "8080",
  "ethsrc": "0A:00:27:00:00:00",
  "protocol": "TCP",
  "source:type": "snort",
  "ip_dst_addr": "192.168.66.121",
  "ip_src_addr": "192.168.66.1",
  "threat:triage:rules:0:score": 10,
  "timestamp": 1504708744000,
  "threat:triage:rules:0:reason": null,
  "threat:triage:score": 10,
  "is_alert": "true",
  "ip_src_port": "50187",
  "guid": "snort_test_2"
}
'
```
### Create a Meta Alert
At this point, we'll group these alerts together.  In Ambari, go to Metron -> Quick Links -> Swagger UI and go to the "Meta Alert Controller".
Click on the create endpoint, and we'll want to send this request (but with the appropriate index subbed in):
```
{
  "groups": [
    "group_one",
    "group_two"
  ],
  "guidToIndices": {
    "snort_test_1":"snort_index_2017.09.06.14",
    "snort_test_2":"snort_index_2017.09.06.14"
  }
}
```
Validate that the metaalert was created and looks good:
```
curl 'node1:9200/m*/_search?pretty'
{
  "took" : 3,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "failed" : 0
  },
  "hits" : {
    "total" : 1,
    "max_score" : 1.0,
    "hits" : [ {
      "_index" : "metaalerts",
      "_type" : "metaalert_doc",
      "_id" : "d414a9f2-7a03-4e47-9108-16b6fc800dd3",
      "_score" : 1.0,
      "_timestamp" : 1504725383439,
      "_source" : {
        "average" : 10.0,
        "min" : 10.0,
        "median" : 10.0,
        "alert" : [ {
          "msg" : "snort test alert",
          "threat:triage:rules:0:reason" : null,
          "ip_dst_port" : "8080",
          "ethsrc" : "0A:00:27:00:00:00",
          "threat:triage:score" : 10,
          "is_alert" : "true",
          "protocol" : "TCP",
          "source:type" : "snort",
          "ip_dst_addr" : "192.168.66.121",
          "ip_src_port" : "50187",
          "guid" : "snort_test_1",
          "ip_src_addr" : "192.168.66.1",
          "threat:triage:rules:0:score" : 10,
          "timestamp" : 1504708744000
        }, {
          "msg" : "snort test alert 2",
          "threat:triage:rules:0:reason" : null,
          "ip_dst_port" : "8080",
          "ethsrc" : "0A:00:27:00:00:00",
          "threat:triage:score" : 10,
          "is_alert" : "true",
          "protocol" : "TCP",
          "source:type" : "snort",
          "ip_dst_addr" : "192.168.66.121",
          "ip_src_port" : "50187",
          "guid" : "snort_test_2",
          "ip_src_addr" : "192.168.66.1",
          "threat:triage:rules:0:score" : 10,
          "timestamp" : 1504708744000
        } ],
        "max" : 10.0,
        "threat:triage:score" : 20.0,
        "count" : 2,
        "guid" : "d414a9f2-7a03-4e47-9108-16b6fc800dd3",
        "groups" : [ "group_one", "group_two" ],
        "sum" : 20.0,
        "status" : "active"
      }
    } ]
  }
}
```
Note that both sub alerts are present, the various counts are filled in, a GUID has been given specifically to this meta alert, etc.

### Searching
Searching from the REST API works mostly as expected.  After the above data has been created, use the search endpoint to run this query:
```
{
  "from": 0,
  "size": 2,
  "indices": [
    "*"
  ],
  "query": "guid=snort_test_2",
  "sort": [
    {
      "field": "_timestamp",
      "sortOrder": "ASC"
    }
  ]
}
```
The result should look similar to the following, and have the messages we'd expect (one from snort and one from metaalerts):
```
{
  "total": 2,
  "results": [
    {
      "id": "snort_test_2",
      "source": {
        "msg": "snort test alert 2",
        "threat:triage:rules:0:reason": null,
        "ip_dst_port": "8080",
        "ethsrc": "0A:00:27:00:00:00",
        "threat:triage:score": 10,
        "is_alert": "true",
        "protocol": "TCP",
        "source:type": "snort",
        "ip_dst_addr": "192.168.66.121",
        "ip_src_port": "50187",
        "guid": "snort_test_2",
        "ip_src_addr": "192.168.66.1",
        "threat:triage:rules:0:score": 10,
        "timestamp": 1504708744000
      },
      "score": 1,
      "index": "snort_index_2017.09.06.14"
    },
    {
      "id": "d414a9f2-7a03-4e47-9108-16b6fc800dd3",
      "source": {
        "average": 10,
        "min": 10,
        "median": 10,
        "alert": [
          {
            "msg": "snort test alert",
            "threat:triage:rules:0:reason": null,
            "ip_dst_port": "8080",
            "ethsrc": "0A:00:27:00:00:00",
            "threat:triage:score": 10,
            "is_alert": "true",
            "protocol": "TCP",
            "source:type": "snort",
            "ip_dst_addr": "192.168.66.121",
            "ip_src_port": "50187",
            "guid": "snort_test_1",
            "ip_src_addr": "192.168.66.1",
            "threat:triage:rules:0:score": 10,
            "timestamp": 1504708744000
          },
          {
            "msg": "snort test alert 2",
            "threat:triage:rules:0:reason": null,
            "ip_dst_port": "8080",
            "ethsrc": "0A:00:27:00:00:00",
            "threat:triage:score": 10,
            "is_alert": "true",
            "protocol": "TCP",
            "source:type": "snort",
            "ip_dst_addr": "192.168.66.121",
            "ip_src_port": "50187",
            "guid": "snort_test_2",
            "ip_src_addr": "192.168.66.1",
            "threat:triage:rules:0:score": 10,
            "timestamp": 1504708744000
          }
        ],
        "max": 10,
        "threat:triage:score": 20,
        "count": 2,
        "guid": "d414a9f2-7a03-4e47-9108-16b6fc800dd3",
        "groups": [
          "group_one",
          "group_two"
        ],
        "sum": 20,
        "status": "active"
      },
      "score": 1,
      "index": "metaalerts"
    }
  ]
}
```

ottobackwards commented:
So, this has me thinking....   It is a shame to have these test materials ( and others like @JonZeolla creates ) embedded in these pr's.

Maybe we should have someplace to put them in the code tree?

mmiklavc commented:
I don't think that's a bad idea. I definitely like having test scripts associated with the PR's because we can explicitly see what was done and/or recommended at the time the PR was being ushered into master. But we could also start to consolidate these into manual testing scripts that we reference ongoing. Then you could say something like:

```
Testing Plan
- es basic search test
- meta alerts search test
- kibana dashboard smoke test
```

That could be links, or we could copy-paste the current manual test plan from the source tree. I think the main thing I would want is to have the PR's test plan (here in the comments) be maintained statically for posterity based on the code at that time. For example, since we're removing the data pruner, the new test plan would not have a data pruner test, but I'd want to make sure any earlier PRs that did have a test plan for it can still be viewed statically. Basically, you'd either include the latest commit as part of the link or copy-paste the test plan.

ottobackwards commented:
Yeah, I'm trying to think of the right way to do it, without having to have it be "officially maintained".
more like an informal /testing_stuff , with some descriptions, when it was valid/written against etc.

Something like that.  Then people could edit them and adapt them etc.

mmiklavc commented:
Just a status update on this. We're currently waiting for 0.4.2 to roll out before this gets committed. We definitely want more eyes and testing on this PR considering its breadth and size. We do not have any +1's yet, and I would prefer to have at least 2 for good measure, if at all possible.

justinleet commented on metron-platform/metron-elasticsearch/README.md:
```diff
@@ -1,5 +1,14 @@
+* [Properties](#properties)
+* [Upgrading to 5.6.2](#upgrading-to-562)
+* [Type Mappings](#type-mappings)
+* [Using Metron with Elasticsearch 5.x](#using-metron-with-elasticsearch-5x)
```
This link is broken, because the section name doesn't line up with the actual name

justinleet commented on metron-platform/metron-elasticsearch/README.md:
```diff
@@ -33,7 +42,217 @@ For instance, an `es.date.format` of `yyyy.MM.dd.HH` would have the consequence
+* [https://www.elastic.co/guide/en/elasticsearch/reference/5.6/breaking_50_mapping_changes.html](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/breaking_50_mapping_changes.html)
+* [https://www.elastic.co/blog/strings-are-dead-long-live-strings](https://www.elastic.co/blog/strings-are-dead-long-live-strings)
+
+## Using Metron with Elasticsearch 5.6.2
```
The ES 2.x nested alert field still exists in this branch, right?  I don't think that requirement got removed.

If not, this section should probably just say it's required in ES, and drop the obsolete portion of the statement..

mmiklavc replied:
All set

justinleet commented on metron-deployment/README.md:
```diff
@@ -1,3 +1,16 @@
+# Metron Deployment
```
Can you please add the license header to this? https://github.com/apache/metron/pull/884 is close to going in and enforcing this, so I'm hoping to avoid impact to master.

```
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->
```

justinleet replied:
As a heads up, #883 is in now, so this will have to be taken care of when you merge master to deconflict.

mmiklavc replied:
All set

justinleet commented on pom.xml:
```diff
@@ -159,7 +159,7 @@
                     <configuration>
                         <!-- Skip the default running of this plug-in (or everything is run twice...see below) -->
-                        <argLine>@{argLine} -Xmx2048m</argLine>
+                        <argLine>-Xmx2048m</argLine>
```
This was in for code coverage via JaCoCo. It basically overrides the empty argline from above that got deleted. Was it causing problems with running things with it there?

mmiklavc replied:
The build was failing with this argLine. Though, I can't recall the exact error now as it's been a while. What's the impact of leaving it off? No code coverage for this module?

justinleet replied:
It's the top level pom, so no code coverage at all, iirc. I haven't looked at it in awhile.  I'll play with it a bit and see what's going on

justinleet replied:
I played with this a bit and I'm not convinced this works in master anymore, so if this was necessary to get this running, I'd rather push fixing coverage off to a follow on task.

mmiklavc commented:
Looking at the additional deployment readme merge conflicts now

mmiklavc commented on metron-deployment/other-examples/manual-install/Manual_Install_CentOS6.md:
```diff
@@ -441,7 +441,7 @@ Client
 - Kibana:
     * Set "kibana_es_url" to `http://<replace_with_elasticsearch_master_hostname>:9200`. "replace_with_elasticsearch_master_hostname" is the IP of the node where you assigned ElasticSearch Master on the Assign Master tab.
-    * Change kibana_default_application to "dashboard/Metron-Dashboard"
+    * Change kibana_default_application to "dashboard/AV-YpDmwdXwc6Ua9Muh9"
```
They changed the links to point to index keys. I was bummed about this as well.

mmiklavc commented:
This is going to take some time to resolve, but everything else in this branch is still unhindered. https://github.com/apache/metron/commit/01c26a77b1041204b0bbbc544cc0a5d02e9339a8#diff-55e8119c8b8ae56260305e01c354d04b

merrimanr commented:
I spun this up in full dev and spent all day testing it.  From a functional perspective, I can not find anything wrong with it.  I ran through the test plan in this PR and everything worked as expected.  I also tested this exhaustively with the Alerts UI and Swagger UI and everything works great.

The only issue I found are with the Alerts UI e2e tests.  These no longer run at all and I suspect it's because the templates have changed with ES 5.6.2 so loading e2e test data and template no longer works.  I'm not sure that this should hold up this PR since e2e tests are actively being worked on in https://github.com/apache/metron/pull/857 but I wanted everyone to be aware.

mmiklavc commented:
Worked with @merrimanr to fix the e2e test issue and just submitted a fix. This does not fix the full e2e test runs as this is being handled by 857, however it brings this PR back to parity with current functionality in master.

merrimanr commented:
I ran this up in full dev again and verified the e2e tests now work similar to how they do in master.  I also manually tested several other areas including the Alerts UI, Kibana and Swagger.  Everything works as expected.  Assuming others are still testing but +1 from me.

justinleet commented:
At this point, I'm +1 since @merrimanr ran up the e2e tests.  A couple people have put a fair amount of testing into this, and it seems like at this point we're at parity and not finding more issues.

cestella commented:
I want to pile on and give this my (non-binding since I contributed PRs against this PR) +1.  LGTM!
